module/data/item/spell.mjs

```diff
@@ -0,0 +1,89 @@
+ * @property {object} components                 General components and tags for this spell.
+ * @property {boolean} components.vocal          Does this spell require vocal components?
+ * @property {boolean} components.somatic        Does this spell require somatic components?
+ * @property {boolean} components.material       Does this spell require material components?
+ * @property {boolean} components.ritual         Can this spell be cast as a ritual?
+ * @property {boolean} components.concentration  Does this spell require concentration?
+ * @property {object} materials                  Details on material components required for this spell.
+ * @property {string} materials.value            Description of the material components required for casting.
+ * @property {boolean} materials.consumed        Are these material components consumed during casting?
+ * @property {number} materials.cost             GP cost for the required components.
+ * @property {number} materials.supply           Quantity of this component available.
+ * @property {object} preparation                Details on how this spell is prepared.
+ * @property {string} preparation.mode           Spell preparation mode as defined in `DND5E.spellPreparationModes`.
+ * @property {boolean} preparation.prepared      Is the spell currently prepared?
+ * @property {object} scaling                    Details on how casting at higher levels affects this spell.
+ * @property {string} scaling.mode               Spell scaling mode as defined in `DND5E.spellScalingModes`.
+ * @property {string} scaling.formula            Dice formula used for scaling.
+ */
+export default class SpellData extends SystemDataModel.mixin(
+  ItemDescriptionTemplate, ActivatedEffectTemplate, ActionTemplate
+) {
+  static defineSchema() {
+    return this.mergeSchema(super.defineSchema(), {
+      level: new foundry.data.fields.NumberField({
+        required: true, integer: true, initial: 1, min: 0, label: "DND5E.SpellLevel"
+      }),
+      school: new foundry.data.fields.StringField({required: true, label: "DND5E.SpellSchool"}),
+      components: new MappingField(new foundry.data.fields.BooleanField(), {
+        required: true, label: "DND5E.SpellComponents",
+        initialKeys: [...Object.keys(CONFIG.DND5E.spellComponents), ...Object.keys(CONFIG.DND5E.spellTags)]
+      }),
+      materials: new foundry.data.fields.SchemaField({
+        value: new foundry.data.fields.StringField({required: true, label: "DND5E.SpellMaterialsDescription"}),
+        consumed: new foundry.data.fields.BooleanField({required: true, label: "DND5E.SpellMaterialsConsumed"}),
+        cost: new foundry.data.fields.NumberField({
+          required: true, initial: 0, min: 0, label: "DND5E.SpellMaterialsCost"
+        }),
+        supply: new foundry.data.fields.NumberField({
+          required: true, initial: 0, min: 0, label: "DND5E.SpellMaterialsSupply"
```

Comment:
`integer: true` maybe?

Reply:
I'm not 100% sure how people are using these fields so I'm reluctant to rule out having decimal values here.

Reply:
Yeah that's fair enough, I don't feel particularly strongly.